Add accessibility guidelines for contributors

thibaudcolas opened the pull request:
As per discussions with @django/accessibility, and [DEP-11](https://github.com/django/deps/pull/69), and  the initial (already-closed) [#31617 "Improve overall admin accessibility"](https://code.djangoproject.com/ticket/31617).

I chose to go with what I think is close to the most minimal set of guidelines possible. In particular,

- This almost only relies on resources from the W3C directly Those are the most authoritative so they should be easy for us to agree on. But in the future we might want to complement them with more contributor-friendly resources.
- The testing guidelines are as minimal as it gets while also attempting covering the most demanding scenarios  – so we don’t overwhelm contributors with lots of checklists.

I see a few issues with this as-is which I chose to leave unresolved for now, so we get something "live" and iterate on it later:

- Accessibility guidance should be dispersed alongside all relevant areas of the contributor docs, rather than being tucked away in the one page.
- This is within the "writing code" sub-section, which sells those guidelines way short of what they can be useful for (and Django "product" contributions more widely?)

It did feel like the right place within the current setup based on how the "[Work on the Django framework](https://docs.djangoproject.com/en/dev/internals/contributing/#work-on-the-django-framework)" docs are described.

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+* Windows Magnifier and macOS Zoom
+* Windows Speech Recognition and macOS Voice Control
+* VoiceOver on iOS, or TalkBack on Android
+* `Contrast themes`_ on Windows
```
This is heavily based on https://docs.wagtail.org/en/stable/contributing/developing.html#accessibility-targets / https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies.

For the more demanding changes we would want to test with _all_ of those things, but is burdensome due to requiring 4 different operating systems. At least there’s something for everyone!

Worth mentioning two possible options I chose to exclude for now:

- [Orca](https://help.gnome.org/users/orca/stable/index.html.en) for Linux. There’s 0 reported usage on [screen reader user surveys](https://webaim.org/projects/screenreadersurvey9/) and 0.5% reported on https://accessibility.blog.gov.uk/2016/11/01/results-of-the-2016-gov-uk-assistive-technology-survey/. I’ve never used it myself nor know anyone who has used it. I think it would be a good one to include _if_ we have anyone contributing in Django who is willing to test with it.
- [ChromeVox](https://support.google.com/chromebook/answer/7031755?hl=en-GB) on ChromeOS. There is some reported usage in survey but as far as I know there is no other way to test it than having a ChromeBook (the browser extension for Chrome isn’t a viable option). Again a good one to include _if_ we have anyone contributing in Django who is willing to test with it.

pauloxnet replied:
> Worth mentioning two possible options I chose to exclude for now:
> * [Orca](https://help.gnome.org/users/orca/stable/index.html.en) for Linux. There’s 0 reported usage on [screen reader user surveys](https://webaim.org/projects/screenreadersurvey9/) and 0.5% reported on https://accessibility.blog.gov.uk/2016/11/01/results-of-the-2016-gov-uk-assistive-technology-survey/. I’ve never used it myself nor know anyone who has used it. I think it would be a good one to include _if_ we have anyone contributing in Django who is willing to test with it.

I believe it is important to also include Orca for Linux in the list of technologies for many reasons, including:
- in the WebAIM survey it is specified that Orca is present among the technologies but has been merged into "others" (see [WebAIM: Screen Reader User Survey #5 Results](https://webaim.org/projects/screenreadersurvey5/#primary) and [WebAIM: Screen Reader User Survey #6 Results](https://webaim.org/projects/screenreadersurvey6/#primary) _"While they were not listed as response options for simplicity's sake, the survey comments indicate that Orca, SuperNova, and Speakup (among others) were common "Other" screen readers."_)
- Linux is the most used operating system among Django developers with 41% (see [Django Developers Survey 2022 Results](https://lp.jetbrains.com/django-developer-survey-2022/#circular-chart-685)), which makes them a decidedly different user base than that of the aforementioned survey, where it is only 1.4% (see [WebAIM: Screen Reader User Survey #9 Results](https://webaim.org/projects/screenreadersurvey9/#os))
- it could be an excellent opportunity to enrich future surveys for Django, Python or even Wgatail by asking actual developers what assistive technologies they use
- I have only tried to use Orca, but if we don't find other users of it in the accessibility team I could help you with this activity

thibaudcolas replied:
Thanks @pauloxnet! Those sound like excellent reasons to me, I’ll add it unless others raise objections.

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Where expert help is available, also consider testing with:
+
+* `JAWS`_ on Windows
+* `Dragon`_ on Windows
```
These feel essential to me because they are very popular, but I chose to keep them separate because:

- Their licenses cost a lot of money
- They’re Windows-only

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+
+Parts of Django do not meet our accessibility targets at the moment.
+We actively work on fixing issues both as part of ongoing maintenance,
+and bigger overhauls. To learn about known issues, and get involved, see:
```
I thought it was important to acknowledge that Django is quite far from meeting all the goals stated above (it’ll be plenty obvious to people going through the Django admin with our stated standards and testing targets in particular) – but that’s something we’re working on (and _you_ can get involved 😉 ).

duranbe replied:
Maybe a link to a board of open issues would be cool, such that possible maintaners can _jump in_

thibaudcolas replied:
There are links to open issues in items 2 and 3 in the list below, is that different from what you’re expecting?

felixxm commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+* 400% browser zoom
+* forced-colors mode DevTools emulation
+
+.. note::
```
Please try to avoid "bare" notes and use `admonition` instead. This guideline is from [Guidelines for reStructuredText files](https://docs.djangoproject.com/en/stable/internals/contributing/writing-documentation/#guidelines-for-restructuredtext-files).

thibaudcolas replied:
Could you say more @felixxm? I chose not to use `admonition` because to me that means the content is a "warning" of sorts. I also didn’t understand that specific guideline as from what I could see the two rendered the same as far as the published docs are concerned.

If this is about the lack of a heading (is that what "bare" means?), I chose not to because though it visually looks like a heading, it’s not properly marked up as such for screen reader users.

---

So if we want to revisit this I think I’d rather just have the content within the main body copy of the docs and avoid admonitions / notes altogether.

felixxm replied:
If it's a warning use `warning`. Notes and admonitions are like side notes, so you theoretically don't need to read them to understand the topic. Titled admonitions are better because you don't have to read the entire note to check what it is about.

thibaudcolas replied:
Should now be addressed.

shangxiao commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Accessibility
+=============
+
+All aspects of Django must be compatible with our accessibility standards,
```
```suggestion
All aspects of Django must be compatible with our accessibility standards
```
As per the other PR I mentioned this with ;)

shangxiao commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+=======================
+
+Django aims for conformance with `WCAG 2.2`_ at the AA level.
+WCAG is the most well established standard for web accessibility. AA-level
```
```suggestion
WCAG is the most established standard for web accessibility. AA-level
```
"most established" is sufficient 👍

thibaudcolas replied:
Thanks! English isn’t my native language so very much appreciate that kind of feedback.

shangxiao commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+
+Django aims for conformance with `WCAG 2.2`_ at the AA level.
+WCAG is the most well established standard for web accessibility. AA-level
+guidelines are the most common legal compliance target around the world.
```
Suggestion: Consider using "worldwide" which is more succinct & common when describing things 'around the world' 😊
```suggestion
guidelines are the most common legal compliance target worldwide.
```

shangxiao commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+----------------------
+
+Where more extensive testing is appropriate,
+here are free assistive technologies we recommend testing with:
```
```suggestion
here are <determiner> free assistive technologies we recommend testing with:
```
We just need a "determiner" here like "some" or "a few"… 👍

shangxiao commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Known issues and how to help
+============================
+
+Parts of Django do not meet our accessibility targets at the moment.
```
Consider prefixing this sentence with "Some" or "There are some", eg:
```suggestion
There are some parts of Django that do not meet our accessibility targets at the moment.
```

shangxiao commented:
Nice to see some a11y guidelines getting added to the docs 👍

I'm not an a11y expert so I've only commented on the grammar where it looks weird to me as a native english speaker.

Side note: I'm a keyboard user and am super quick to notice when sites violate the keyboard-friendly parts of a11y… eg my pet hate is when devs take shortcuts by making div based buttons 🤯  I wonder if a note about how good a11y can benefit everyone? 🤔

thibaudcolas commented:
Thanks for the review @shangxiao, much appreciated! That’s a great point re accessibility benefitting everyone, I’ll try to think a bit of where to add it in.

knyghty commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Accessibility standards
+=======================
+
+Django aims for conformance with `WCAG 2.2`_ at the AA level.
```
"aims" feels like a bit of a cop-out, I'd prefer stronger language here.

thibaudcolas replied:
I’ll try!

Scotchester replied:
There's a tension that's hard to rectify: We definitely want to use strong language, but we also have to be honest about the fact that we're not there yet. In my suggestion on the intro above, I chose "We work … to …" as wording that indicates we are actively doing something, not just aiming/aspiring/seeking. Perhaps that sort of change would be good here, as well.

thibaudcolas replied:
I’ve tried to make this stronger, without making it sound like we’re there yet.

knyghty commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+a wide range of input devices and assistive technologies.
+The best way to do this effectively is to take accessibility considerations
+into account as part of designing features. If in doubt, consult
+with users who rely on assistive technologies or with accessibility experts.
```
> in the spirit of a "shift-left" approach.

What happened to not using technical jargon in documentation, Thibaud? 🤔

knyghty replied:
But also, all good points.

thibaudcolas replied:
😂 you got me

nessita replied:
I agree with the paragraph but as I mentioned in a PR-level comment, I think it's scary. I believe it should have links to concrete resources to quantify how many are enough "accessibility considerations" and who/where to reach out for help.

thibaudcolas replied:
Also noting [Andrew’s comment here](https://github.com/django/django/pull/17338#issuecomment-1843410088) as I believe this will be the right place to address this (TBC):

> As for the guidelines around strictness - I think it is especially important to say the implied part out loud (that progressive improvement in small parts is better than demanding perfection). We don't want to discourage people from contributing, we want to encourage them to instead make things a little bit better every time.

This was a reply to [my prev comment](https://github.com/django/django/pull/17338#issuecomment-1842250001)

knyghty commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+* Automated checks in the `Accessibility Insights`_ browser extension,
+  or an equivalent tool with the `Axe`_ checker.
+
+Where appropriate, also test with:
```
I think we need to define (or at least give examples) when it's appropriate.

thibaudcolas replied:
Great idea, will do.

nessita replied:
Big +1, otherwise the "when" is subject to interpretation and subsequent potential frustration.

knyghty commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+We actively work on fixing issues both as part of ongoing maintenance,
+and bigger overhauls. To learn about known issues, and get involved, see:
+
+* ``#accessibility`` on the `Django Discord server`_.
```
Nice.

knyghty reviewed (commented):
I think this is a great start.

One thing I want to mention is that I feel it's important to strike a balance between encouraging accessibility testing and making contributing easier as these seem like slightly conflicting goals. I think we should make it very explicit that if you are unsure about accessibility testing it's fine to - I'm not sure but e.g. tag the accessibility team on the PR and we can take a look for them.

I don't want these well-intentioned changes to deter anyone from contributing - especially to the admin which needs some love - because of the extra burden this adds.

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Accessibility
+=============
+
+All aspects of Django must be compatible with our accessibility standards,
+and with the specific support targets we aim for.
```
I'm not sure about the suggested use of "The Django project" here. Perhaps just saying "Django is committed …" would be better, but I'm not well-versed in the established practice for how we refer to ourselves.

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Accessibility standards
+=======================
+
+Django aims for conformance with `WCAG 2.2`_ at the AA level.
```
We should spell out this acronym on first usage.

```suggestion
Django aims for conformance with the `Web Content Accessibility Guidelines (WCAG), version 2.2`_, at the AA level.
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+WCAG is the most well established standard for web accessibility. AA-level
+guidelines are the most common legal compliance target around the world.
+
+Wherever applicable, we also attempt to follow other best practices, such as:
```
Similar to Tom's comment on "Where possible", the "Wherever applicable" here could benefit from some elaboration. Or maybe it should just be dropped:

```suggestion
We also attempt to follow other best practices, such as:
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+
+* WCAG 2.2 AAA-level guidelines.
+* The upcoming `WCAG 3.0`_ guidelines.
+* `ATAG 2.0`_ guidelines for authoring tools.
```
```suggestion
* `Authoring Tool Accessibility Guidelines (ATAG) 2.0`_  – guidelines for software and services that are used to produce web content.
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+    standards? We recommend `The A11Y Project`_,
+    a community-driven effort to make digital accessibility easier.
+
+.. _WCAG 2.2: https://www.w3.org/TR/WCAG22/
+.. _WCAG 3.0: https://www.w3.org/TR/wcag-3.0/
+.. _ATAG 2.0: https://www.w3.org/TR/ATAG20/
```
```suggestion
.. _Web Content Accessibility Guidelines (WCAG), version 2.2: https://www.w3.org/TR/WCAG22/
.. _WCAG 3.0: https://www.w3.org/TR/wcag-3.0/
.. _Authoring Tool Accessibility Guidelines (ATAG) 2.0: https://www.w3.org/TR/ATAG20/
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Accessibility testing baseline
+------------------------------
+
+ Always test user interface changes with:
```
Space-o:

```suggestion
Always test user interface changes with:
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+ Always test user interface changes with:
+
+* Keyboard-only navigation
+* Automated checks in the `Accessibility Insights`_ browser extension,
+  or an equivalent tool with the `Axe`_ checker.
```
Maybe slightly clearer?

```suggestion
* The `Accessibility Insights`_ browser extension's automated checks feature,
  or an equivalent tool with the `Axe`_ checker.
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+
+* Touch-only navigation
+* 400% browser zoom
+* forced-colors mode DevTools emulation
```
Suggest we name-check the most common use case, and also link to instructions for how to emulate.

```suggestion
* Forced-colors mode (e.g., Windows Contrast Themes)
  * `Can be emulated with Chrome/Edge dev tools`_
  
.. _Can be emulated with Chrome/Edge dev tools: https://devtoolstips.org/tips/en/emulate-forced-colors/
```

Scotchester commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+.. note::
+    How much should you test? Devise the UI with accessibility in mind,
+    and the testing will only be needed as a final check. For more complex
+    interfaces, confer with other contributors to decide on testing targets.
```
I agree that it's a worthwhile thing to include.

Scotchester reviewed (commented):
Excellent work! A few suggestions for your consideration.

thibaudcolas commented:
This is now ready for another round of review. Aside from specific edits, I have:

- Changed the order of _Accessibility testing baseline_, so it’s clearer how/much, when to test
- Added a call to reach out to the accessibility team for help
- Added a reference on how accessibility improvements benefit everyone
- Removed admonitions.

thibaudcolas commented:
I believe we need feedback on those guidelines from a wider group than we have currently – so have created a [forum thread](https://forum.djangoproject.com/t/feedback-needed-accessibility-guidelines-for-contributors/25437) and [tweeted about this](https://twitter.com/thibaud_colas/status/1725847406858719387).

The feedback so far has been excellent, but I’d like to hear from folks who haven’t been working in open source or on accessibility and whose contributions might be most impacted by those guidelines 💥🌈. Please share the forum thread or the tweet as you see fit!

duranbe commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+  for software and services that are used to produce web content.
+
+To learn more about accessibility without diving straight into standards,
+we recommend `The A11Y Project`_, a community-driven effort
```
An acronym definition of A11Y would be good here imo

thibaudcolas replied:
Hmmm interesting. Since this is the name of the website as they write it, I was hesitant to spell it out.

carltongibson commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+
+Always test user interface changes with:
+
+* Keyboard-only navigation
```
Is there some resource that we could link to that gives a checklist for what to cover here? 

I can tab about, etc., but likely there’s lots that I wouldn’t think or know consider.

thibaudcolas replied:
Good point, yes definitely! The TL;DR; is you should be able to do everything with just tab, arrow keys, space and enter. There are some more nuances which a resource will be helpful for.

thibaudcolas commented:
Noting feedback I have received since the last call for feedback:

> It might be kind of daunting for new contributors who don't know much (or anything!) about accessibility; maybe it could be helpful to have a beginner-friendly link (maybe the one for [The A11Y Project](https://www.a11yproject.com/) that's already linked in the doc) as part of the very first paragraph, or at least before all the links for the different standards. It might make it easier for an accessibility beginner to get through the docs without them seeming too overwhelming

it seems tricky to me to add a link to the first paragraph as-is, and we’ve already spent lots of time refining its text, but we could certainly have a beginner-friendly link first before the one about standards

> maybe it's also missing a section on how to unit test accessibility improvements / fixes? I'm really used to frontend unit testing (in JS, e.g with testing-library) and found it quite challenging to successfully add a test in my PR to Django

Agree 100% with this but I’m not sure what to do about it, as the main way I’d normally recommend writing those tests isn’t available in Django. It also feels like a specific enough point that it could be worth a separate PR or forum thread to discuss QA options.

carltongibson commented:
@thibaudcolas, maybe a callout (.. admonition, or similar) for the first point... 🤔

> Accessibility is a big topic...

andrewgodwin commented:
I see a Steering Council review was requested on this - I'm not enough of an accessibility expert to really review the content in deep detail, but is it just approval of the overall direction and tone that's wanted?

nessita commented:
Hey @thibaudcolas, I read the HTML guidelines. Thank you for working on this!

My feedback is based on the fact that I'm not an expert in accessibility and that many of the things I've seen or learn are due to doing and reading reviews in the Django project. So with that spirit, my first reaction to this guideline proposal was "this is great stuff but I have no idea how to do it or how to comply with this".

To that extent, I think it would be super important to accompany this contribution guide with a handful of How-Tos. And this guide could link to specific a How-To depending on the guide context. Has this been considered?

thibaudcolas commented:
@andrewgodwin ah thanks Andrew, that’s probably my bad for not understanding the remit of the council. Essentially I was trying to figure of all the "senior contributor" groups in Django land, which one(s) would be a good fit to check the guidelines will be good enough for future contributors. I referred to [how the steering council is described in docs](https://docs.djangoproject.com/en/dev/internals/organization/#steering-council) and descriptions of the [technical advisory team](https://www.djangoproject.com/foundation/teams/#technical-advisory-team-team) and [technical team](https://www.djangoproject.com/foundation/teams/#technical-team-team) to figure this out.

I went for the steering council because of the "_Their main concern is to maintain the quality and stability of the Django Web Framework._", which this is very much about, but can also take this elsewhere if you think it makes more sense? Perhaps you could help us by defining what is "enough reviewing" for something like this?

> The one thing I would want us to have a position on, as a project, is where the bar is for turning down features or PRs because they are not accessible - how strict are we? I personally think it's fine if they could be made accessible without big code changes but don't come fully accessible due to the contributor's time or skillset, but that is not very clear to me reading it right now.

Interesting, do you think that’s something for those guidelines to cover? From my perspective this feels like something where we would be pragmatic. There are lots of issues in the UI right now, so I’d say it’s fine if new work comes with fewer "accessibility issues per line of code" than there are right now. But gradually as the quality gets higher, we’d want to be stricter.

Is this kind of sliding scale of quality used elsewhere in Django contributions either formally or informally for similar things where contributor expertise varies a lot? For example amount of docs, performance, security, amount or type of tests, etc

---

@nessita thanks for the feedback ☺️ we’ve discussed adding more guidance / linking to more resources. I think people are generally keen. I was hesitant initially (because of the work involved), but I think it’s a great point that it’ll be invaluable to people who are new to this, so this feels even more important for us to do.

Could you comment on specific points of the proposed guidelines you’d think need a "how-to"? As far as the work involved – there’s lots of great resources out there. We’d need to review a few options, and have enough agreement of which ones are high-quality enough to be referenced in the Django docs.

Does Django have any kind of policy on what is allowed or not as far as linking to external sites?

nessita commented:
> Could you comment on specific points of the proposed guidelines you’d think need a "how-to"? As far as the work involved – there’s lots of great resources out there. We’d need to review a few options, and have enough agreement of which ones are high-quality enough to be referenced in the Django docs.
> 
> Do we have any kind of policy on what is allowed or not as far as linking to external sites?

As far as I know, we prefer not to link to external sites unless they are super well known and stable (like python.org).

My initial thinking on providing How-Tos was that we would write and provide those ourselves, not linking to external sites. Specifically, I was hoping we would add new pages to our "How-to" guides (https://docs.djangoproject.com/en/5.0/howto/). A few I can think of:

* How to use keyboard navigation (in the admin and perhaps "in your site")
* How to (and when) properly use the aria* attributes in widgets and other parts of a site. Perhaps mention screen readers here? 
* How to user voice control systems, with a focus (perhaps?) on non Apple devices

Then, the accessibility guide proposed proposed in this PR has, IMHO, some "scary to newbies parts" such as:

> The best way to do this effectively is to take accessibility considerations into account as part of designing features. If in doubt, consult with users who rely on assistive technologies or with accessibility experts.

As a contributor with little or no experience with accessibility, reading that paragraph may make then feel like they should know beforehand what are the "accessibility considerations", and that if they don't, they should consult with users/experts. But who are those users and experts? Where are they? How can they reach them? It feels a little like we are making them do (more) homework and I fear that will discourage them. I would suggest that we rework this text to:

1. Link "accessibility considerations" to a list (built by us) about the actual considerations we expect to be covered
2. Provide ways of contact the groups of users and experts. Hopefully this is just a link to the accessibility Discord channel or to a Forum category

I'll add a few more suggestions/comments as PR comments. Thanks!

andrewgodwin commented:
@thibaudcolas I believe the problem is that there is no "senior contributor" group after the dissolution of Django Core, rather deliberately in that case. I'm not sure I agree with it entirely, but that was the intention, and so you'll basically just have to rely on whatever consensus you can get from whoever is interested.

I agree the Steering Council/Technical Board descriptions imply that it might do this, but it has not and DEP-10 does not provide for it to.

As for the guidelines around strictness - I think it is especially important to say the implied part out loud (that progressive improvement in small parts is better than demanding perfection). We don't want to discourage people from contributing, we want to encourage them to instead make things a little bit better every time.

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+
+.. _accessibility-support-targets:
+
+Accessibility support targets
```
```suggestion
Support scope
```

thibaudcolas replied:
Could you explain your suggestion? I have spent a lot of time on wording choices here, and based this on other existing guidelines that use this wording.

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+
+.. _accessibility-testing-baseline:
+
+Accessibility testing baseline
```
```suggestion
Testing baseline
```

nessita replied:
Even if someone lands in a heading, the topic of the document is "Accessibility". Having this prefixed on every section title feels redundant and I think it's fair to expect people to be able to scroll up and down for context if they need it.

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+Design the UI with accessibility in mind,
+and the testing will only be needed as a final check. For more complex
+interfaces, confer with other contributors to decide on testing targets.
+Reach out to the `Accessibility team`_ for support and to coordinate testing
```
Remember the wrapping at 79 cols :-)

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+
+* Touch-only navigation
+* 400% browser zoom
+* Forced-colors mode (for example Windows Contrast Themes)
```
I would love to see each one of these be a link to a How-To provided by us.

nessita replied:
Just to clarify my point here, I think we need to describe somewhere how to achieve forced-colors mode (I'm honestly asking out of ignorance). Is it a browser setting? Is it browser specific? Is it an operating system setting?

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+Assistive technologies
+----------------------
+
+Where more extensive testing is appropriate,
```
Going back to @knyghty's point, we should make explicit where is appropriate.

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+Where more extensive testing is appropriate,
+here are popular free assistive technologies we recommend testing with:
+
+* `NVDA`_ on Windows with Firefox ESR
```
Does this mean that these tools are *not* available for non-ESR Firefox?

thibaudcolas replied:
No, it just means we recommend testing with Firefox ESR.

nessita commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+* VoiceOver on iOS, or TalkBack on Android
+* `Contrast themes`_ on Windows
+
+Where expert help is available, also consider testing with:
```
I'm not sure what this means :thinking:

thibaudcolas replied:
I’ll rephrase. This means "If you have the expertise or you can work with someone who does, also test with:".

nessita reviewed (commented):
Amazing effort, thank you!!!

I think this is necessary adding to our docs, though I would advocate to only merge it when the concrete How-Tos are available.

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,123 @@
+Accessibility
+=============
+
+The Django project is committed to being accessible to everyone
+in all possible ways.
+We work to conform to established accessibility standards,
+and also to meet our own targets for supporting specific input devices
+and assistive technologies.
```
I’ll rewrite this whole document to wrap exactly at 79 columns for every line. For the record, I find this really problematic for readability of the documentation, so in the future I’ll raise an issue to change this rule (probably after #17340).

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+Wherever applicable, we also attempt to follow other best practices, such as:
+
+* WCAG 2.2 AAA-level guidelines.
+* The upcoming `WCAG 3.0`_ guidelines.
```
Update – I’m going to remove this unless others disagree, for the purpose of making this content more approachable (no concerns about there being multiple versions of WCAG to be aware of). WCAG 3.0 feels like it’s at least 5 to 10 years away at a minimum, we can add it back once it’s closer to ready.

thibaudcolas commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,114 @@
+
+* Keyboard-only navigation
+* Automated checks in the `Accessibility Insights`_ browser extension,
+  or an equivalent tool with the `Axe`_ checker.
```
@nessita the wording as-is says "Accessibility Insights or equivalent tool with the Axe checker", so I don’t understand how you equate this to forcing Chrome on contributors? There are a lot of tools available with this checker (a browser extension for Firefox, lots of command line tools, etc).

Do you think we should link to specific options here then? Perhaps link to a list of products / integrations of Axe?

nessita replied:
I guess my point is that, upon reading this sentence, I'm concerned that this approach might be intimidating for new contributors, as it places the burden of finding an "equivalent tool with the Axe checker" squarely on their shoulders.  Does this make sense?

sarahboyce replied:
The new contributor experience is tough and I get your point here - but I think this is fine

My reasons are that given roughly 65% of browser users use a chrome based browser, having an expectation that someone will test against a chrome based browser when making UI updates is (to me) not unreasonable
Our selenium tests also only run against chrome in the CI and (as a result) have some failures on firefox, so the chrome preference is an existing part of the Django contributing process.

In terms of the new contributor experience, the main message is roughly "no one makes a contribution to Django on their own, if you need help - ask". But I think that is true of all contributions and the start of this section says `Reach out to the Accessibility team for support and to coordinate testing.`. So I think we have done a reasonable effort to say this.
As we request the accessibility team to review PRs (and given these new guidelines we should be able to feel we could do this if we wanted), "someone" should have done this check (ideally the contributor and the reviewer but at least a reviewer)

As a future TODO, we can add some of this in the PR template around the UI changes (we currently tell them to add screenshots - but we can say "and test for...")

thibaudcolas commented:
Thanks @andrewgodwin, based on your comment I feel like we have enough senior contributors involved as far as consensus then (assuming people who have already reviewed will be happy to re-review further down the line).

I’ll address the point about strictness of the guidelines directly in there in a future draft.

---

@nessita thank you for the feedback 😌. I’ve replied to specific points I’m not clear on. For how-to content, I agree this is important material for us to make available but I have issues with the approach you suggested – specifically that it needs to be content we write and maintain in the Django docs, and that you recommend only merging the guidelines when this content is available. My concerns are that:

- At the current pace, having to write this content as a precondition to adopting those guidelines, will delay adoption of the guidelines by multiple months. We have a very small group of people who would be able to research, author, and review this content.
- There isn’t that much specific to Django to warrant writing "our own" material. The one thing I could think of is perhaps clearly sign-posting all the known issues so people don’t spend their time testing those, but that feels more like "working document" material than "how-to in the docs".
- We’re not that good at accessibility compared to some of the more established experts out there. There will be much better resources out there than what we can provide, and it’ll be simpler to keep our guidelines up-to-date if we can switch references over time as needed, rather than having to write our own material always.

On the specific three you suggested,

> How to use keyboard navigation (in the admin and perhaps "in your site")

I’d like us to have a how-to about that but there’s nothing specific to Django here aside from warning people about current known issues. In terms of writing effort, for me this would be on the order of 5-10 hours of work for a good first draft (so about 2-3 months depending on how much of my "Django contributions" time I spend on this). Finding the one best link is probably closer to 1h of work.

> How to (and when) properly use the aria* attributes in widgets and other parts of a site. Perhaps mention screen readers here?

Again for a very minimal version of this, we’re talking at least 10 hours of work (so 2-3 months for me) to produce something that won’t cover any of the specifics of Django. A how-to with Django specifics (templates, context, multilingual support) would be on the order of 20-30 hours (so I’d say 6 months if it was me writing it).

> How to user voice control systems, with a focus (perhaps?) on non Apple devices

Personally I’ve never used voice control systems on anything but Apple devices. I’m not even aware of what voice control systems are available elsewhere, aside from Dragon on Windows which costs a fortune. Such a guide would definitely require covering Dragon, so we’re talking 30-50 hours of work for someone like me to get up to speed, finding funding for a budget of about $1000. I can’t really recommend we’d invest time on this but if we did I’d say it’s also on the order of 6 months to a year of work at the current pace.

---

So with the team that we have currently to work on this, I’d recommend we spend our time focusing on more minimal guidelines and extend those guidelines with external resources where there’s a need. If people want to write how-tos in the future that sounds great to me, though I’d rather those people contributed to accessibility efforts in other ways.

thibaudcolas commented:
@django/accessibility discussed this point about how-tos in our last meeting. We’re keen to proceed with references to _external_ how-tos in the interest of time / ROI for the reasons I outlined above. @nessita could you confirm whether that seems ok to you?

If people see a need for more Django-specific resources, we thought we could produce material on a theme of "top 10 things to test in Django for <specific accessibility consideration>". For example, we could produce more documentation for components with complex keyboard interactions. Or document correct and incorrect patterns of color use in Django.

thibaudcolas commented:
@nessita 👋 this is waiting for your reply. Could you either take a look, or delegate this to someone else if you don’t have the time?

In the meantime I have updated [the guidelines’ HTML preview](https://django-docs-preview.netlify.app/internals/contributing/writing-code/accessibility).

carltongibson commented:
> Could you either take a look, or delegate this to someone else if you don’t have the time?

Hey @thibaudcolas — the Fellows have a much higher workload than fits in the available. Please be patient, and understand that pressure for an early review is unhelpful, and is a cause of stress. (There's plenty of that coming from folks who aren't high-visibility members of the community.) It's almost certain that @nessita has looked at this several times, and is just trying to find the mental space needed to reply appropriately. I understand you'll be keen to push this forwards, but she'll/they'll get to it in their time. Thanks 🙏

thibaudcolas commented:
Thank you 🙏 I re-raised this because (to me) this is the top priority on the accessibility team’s list. Having this place would help us in delegating reviews to people outside the team / just be faster at reviewing contributors’ pull requests and bug reports like https://code.djangoproject.com/ticket/35115.

@nessita sorry if I’m annoying by reaching out on Slack and here – I just commented here because I hadn’t updated the PR when I raised this on Slack.

I appreciate time is short for everyone but all we need here is a reply on "are how-tos really a must to merge this or not". If it’s yes, then we’ll figure out how to get the team working on those how-tos over 2024, and in the meantime publish those guidelines elsewhere. If it’s "no", then I can proceed with asking others here for review.

nessita commented:
> Thank you 🙏 I re-raised this because (to me) this is the top priority on the accessibility team’s list. Having this in place would help us in delegating reviews to people outside the team / just be faster at reviewing contributors’ pull requests and bug reports like https://code.djangoproject.com/ticket/35115.

Understood, thanks for the extra rationale.

> I appreciate time is short for everyone but all we need here is a reply on "are how-tos really a must to merge this or not". If it’s yes, then we’ll figure out how to get the team working on those how-tos over 2024, and in the meantime publish those guidelines elsewhere. If it’s "no", then I can proceed with asking others here for review.

That's the conundrum I face: my current response is *I'm not sure*. I do believe that internal how-tos are necessary, perhaps starting with a simple task, such as building a basic but accessible page using Django. However, given your strong stance that how-tos should link to external resources, I feel it's essential to invest time in reviewing those external resources. This way, I can better evaluate whether they fulfill the needs of Django users and contributors before potentially contradicting that approach.

Would you please share a short list of the external how-tos that were considered for linking from the Django docs?

thibaudcolas commented:
Thanks! We haven’t considered any specific ones so far, but here are my go-to authoritative resources:

- [A11Y Project – Checklist](https://www.a11yproject.com/checklist/). I consider it authoritative because it’s an independent project dedicated to accessibility, not affiliated with any particular vendor.
- Anything from the W3C Web Accessibility Initiative (WAI)’s Education & Outreach WG. For example [https://www.w3.org/WAI/test-evaluate/preliminary/](https://www.w3.org/WAI/test-evaluate/preliminary/). Possibly a bit dated but as authoritative as it gets.
- [MagentA11y](https://www.magentaa11y.com/how-to-test/) for more complete information. Less authoritative (it’s by T-Mobile), but very comprehensive and high-quality

I expect we’d want to pick resources that work well for beginners, so if the external ones seem reasonable to you I might ask beginners to review them and give us their picks? Perhaps with a pre-triage by the accessibility team?

For example here are the Keyboard resources:

- [A11Y Project checklist – Keyboard](https://www.a11yproject.com/checklist/#keyboard)
- [WAI Easy checks – Keyboard](https://www.w3.org/WAI/test-evaluate/preliminary/#interaction)
- [MagentA11y – Keyboard](https://www.magentaa11y.com/how-to-test/keyboard-focus/)

Perhaps others can share their thoughts too on what kind of resources would be most helpful in this context?

thibaudcolas commented:
Just reflecting on this a bit more – the docs suggested in #17340 would make for solid how-tos if we wanted to invest into our own. Everything on there is specific to Django.

nessita commented:
> Just reflecting on this a bit more – the docs suggested in #17340 would make for solid how-tos if we wanted to invest into our own. Everything on there is specific to Django.

Hey @thibaudcolas, I'm a bit confused by this message, could you please expand a bit your thinking here? What docs specifically would be solid how-tos?

Just in case the following is not widely known, the docs we add would ideally follow the [Diataxis](https://diataxis.fr/) approach.

thibaudcolas commented:
@nessita I would say close to everything proposed in #17340. So for example:

- ImageField and alternative text in Django
- Accessible markup for Django firms
- Semantic HTML pagination with `Paginator`
- Outputting tagged PDFs

nessita commented:
> @nessita what do you think is the path forward here? I think it's too onerous for us (at least for the moment) to maintain a lot of a11y documentation ourselves. 

Firstly, I'd appreciate hearing/reading others' views on this; I'd rather not be the sole decision-maker on whether to land this or not.

Secondly, regarding my ask of "how-tos", I think I could provide some clarification to ensure we are talking about the same thing. For example, I read the links shared by @thibaudcolas, specifically the ones about keyboard navigation. The first one says:
![image](https://github.com/django/django/assets/124304/349dc49c-0b94-445a-869a-574286773f84)

My worry is that, while the text emphasizes the importance of keyboard navigation, it doesn't provide practical guidance on how to achieve it. If a contributor proposes a PR and identifies that keyboard navigation is lacking, there should be clear instructions on what steps the contributor needs to take to address this. Essentially, if we're advocating for accessibility in PRs, my instinct is that we should also include explicit guidance on how to implement it (and likely the intro paragraph in this doc would reference the well known external sources describing the goal). Does this make sense?

> Even if we were to maintain our own guides in the future (which would likely require expanding the a11y team considerably) I don't think these should block this PR. What do you think?

I'm doubting if by guides we are talking about the same thing. I agree we should not (and can not possibly) maintain a whole set of docs about accessibility theory, but I think it would be beneficial to have some guides on how to achieve *concrete things in Django* related to a11y.

A not accurate but perhaps helpful reference would be our CSRF docs, they do not explain the attack in detail, they reference well known sites (owasp, RFCs) if the user wants to know more, but we do provide a "How to use Django’s CSRF protection" to actually mitigate the attack in a Django site.

In summary, if there are well known sites providing concrete/practical guidance on how to do the things we require, I think we could definitely make a compromise and push this forward.

thibaudcolas commented:
@nessita thank you for looking into this again, I’m still not clear if you’re talking about how-tos for Django contributors, or for developers using Django? I agree there’s value in providing accessibility how-tos for Django users like the [How to use Django’s CSRF protection](https://docs.djangoproject.com/en/5.0/howto/csrf/) you’re referring to, but that’s not really the point of this PR which is about guidelines for contributors.

If we’re talking about how-tos for contributors – I think we need to keep in mind here that currently we offer close to zero  information for them to identify what’s important as far as accessibility. You mention the need for:

> clear instructions on what steps the contributor needs to take

It seems to me like we have this already in the current draft ("conform to WCAG 2.2 AA", "confer with other contributors", "reach out to the Accessibility team for support"). 

There’s certainly room for us to link to more guidance for contributors with no accessibility experience working on their own. Contrary to what you said I think [A11Y Project checklist – Keyboard](https://www.a11yproject.com/checklist/#keyboard) is as practical as it gets, but if that’s not good enough we can spend more time working with our contributors to find better resources. I’d just like that process to not hold up what’s already pretty reasonable guidelines.

nessita commented:
> @nessita thank you for looking into this again, I’m still not clear if you’re talking about how-tos for Django contributors, or for developers using Django? I agree there’s value in providing accessibility how-tos for Django users like [How to use Django’s CSRF protection](https://docs.djangoproject.com/en/5.0/howto/csrf/), but that’s not really the point of this PR.

I'm talking about how-tos (or similar, lightweight pages) for Django contributors. I only used the CSRF docs as an example of what I meant with describing **theory vs practical** guidance:

* Theory links to external resources (RFC and related), but
* Practical guidance was provided in simple and straightforward terms, and can be applied even if the reader has not read nor understand the theory in depth

> If we’re talking about how-tos for contributors – I think we need to keep in mind here that currently we offer close to zero information for them to identify what’s important as far as accessibility.

I agree. And this PR provides the guide for knowing what's important (I think there is no discussion about that).
What I think is missing, is explaining *how to achieve* the desired results.

>  You mention the need for:
> > clear instructions on what steps the contributor needs to take
> 
> It seems to me like we have this already in the current draft ("conform to WCAG 2.2 AA", "confer with other contributors", "reach out to the Accessibility team for support").

While the current draft mentions "_conform to WCAG 2.2 AA_", "_confer with other contributors_", "_reach out to the Accessibility team for support_" there's a nuanced perspective on this. Requesting someone who is entirely new to accessibility to "conform to WCAG 2.2 AA" might feel like a daunting task. As someone who started as an accessibility novice, I can attest to the overwhelming and intimidating nature of such a request, particularly because the magnitude of the task is unclear. It might be beneficial to consider more approachable and gradual steps for contributors new to accessibility, ensuring a smoother onboarding process.

> There’s certainly room for us to link to more guidance for contributors with no accessibility experience working on their own. Contrary to what you said I think [A11Y Project checklist – Keyboard](https://www.a11yproject.com/checklist/#keyboard) is as practical as it gets

The link is very practical in terms of **checking** keyboard navigation, but it does not say how to fix something that does not allow for keyboard navigation.

In other words: as a contributor, I open the link, I read "_Can a person navigating with a keyboard or screen reader move around the page in a predictable way?_", I try it, I can't do any useful nor predictive keyboard navigation, so the check fails. How can I fix it? That's the part that I think is missing (but necessary).

> I’d just like that process to not hold up what’s already pretty reasonable guidelines.

While I understand the importance of not delaying the landing of these (already reasonable) guidelines, I think I have a valid concern about the potential intimidation factor for contributors, particularly those new to the topic. Do you agree with that?

carltongibson commented:
I've been following the discussion here. I think **some** _How-to(s)_ is/are needed

I still look at the proposal and am at a loss as to _What it means for me as a PR author or reviewer_. 

There are a lot of links, which go to quite scary documents, from the sample I clicked. It feels like I have an **awful lot to learn** if I'm going to comply with the new guidelines. That's scary. 

> I think it's too onerous for us (at least for the moment) to maintain a lot of a11y documentation ourselves. 

OK. But the must be some basics that cover the 80:20 and are to some extent eternal? Is that right? 

If so, could some worked examples/How-tos for those help? Perhaps with links to other resources (with a _last updated_ marker)? 

The example from one of the reviews above was _Keyboard navigation_. OK, users should be able to do that. But what exactly counts as success there? I use tab a bit. I'm guessing though keyboard navigators do a bit more than that. But what? No idea. Not sure how to learn. (So not sure how to comply with the guidelines here.) 

> ...some basics that cover the 80:20...

Does that exist? Would that be doable?

thibaudcolas commented:
> I agree. And this PR provides the guide for knowing what's important (I think there is no discussion about that).
> What I think is missing, is explaining how to achieve the desired results.

In my opinion this is out of scope for accessibility guidelines. we don’t have much capacity to provide guidelines for _testing_ currently, we certainly don’t have capacity to provide guidelines for implementation. And there is not much special about Django that would warrant having our own "how to implement keyboard support" documentation.

>  I think I have a valid concern about the potential intimidation factor for contributors, particularly those new to the topic. Do you agree with that?

It is a very valid concern but I’d like to think we’ve done a reasonable job of helping with this by providing the right links. We’re recommending the A11Y Project and the Accessibility Insights browser extension specifically because they are beginner-friendly.

---

@carltongibson

thibaudcolas commented:
I think we’re letting perfect be the enemy of good here :/ We have no accessibility guidelines for Django at all, even if just to state "Django aims to conform to WCAG 2.2 AA", which is the fundamental information that’s relevant to all contributors. Friendliness to new contributors is certainly a consideration, but I can’t imagine the status quo of "no docs at all" being any more friendly than what we’re proposing here.

> I agree. And this PR provides the guide for knowing what's important (I think there is no discussion about that).
> What I think is missing, is explaining how to achieve the desired results.

In my opinion this is out of scope for Django accessibility guidelines. There is not much special about Django that would warrant having our own "how to implement keyboard support" documentation.

>  I think I have a valid concern about the potential intimidation factor for contributors, particularly those new to the topic. Do you agree with that?

It is a very valid concern but I’d like to think we’ve done a reasonable job of helping with this by providing the right links. We’re recommending the A11Y Project and the Accessibility Insights browser extension specifically because they are beginner-friendly.

---

@carltongibson yes, it’s a lot to learn. I think it’s reasonable for us to expect that people interested in contributing to UI in Django will either take the time to learn this, or defer to more experienced people, as appropriate. 

> The example from one of the reviews above was Keyboard navigation. OK, users should be able to do that. But what exactly counts as success there?

We can certainly provide a link to test for keyboard navigation if we think that helps. I still think that the contributing docs stating "please make sure that keyboard navigation works" is better than the status quo.

carltongibson commented:
I'm confused. 🤔

Not a [couple of comments ago](https://github.com/django/django/pull/17338#issuecomment-1894539190) you had a list of your go to resources, followed (a comment or two later) by some examples of good Django-specific cases where they could be shown in action. 

A how-to/tutorial/something showing what accessibility means in the context of a Django project, and how to check for that, on a PR to Django, or equally when you're working on your own project, would go miles here. And AFAICS is all that's needed. 

It's like any part of the contributing process: we have to document how it's done, so contributors (new and old) know what to do. 

> Make sure you run the linters!

OK, here are the docs on [setting up pre-commit](https://docs.djangoproject.com/en/5.0/internals/contributing/writing-code/coding-style/#coding-style-pre-commit). 

> Make sure you follow the Accessibility Guidelines. 

Erm... — We need something to say here. 

The contribution docs aren't "perfect", but we can't just abandon folks here. (That wouldn't be "good" either.)

I'm all for this endeavour but I read the PR again yesterday, and my take-away was that I wouldn't have clue where to begin. I'd have to spend a lot of time reading the linked resources, which is something I can do, but we need a **Start Here!**. 

(Or so I think.)

thibaudcolas commented:
Alright, I’ll do another draft where there’s a "start here" section at the top then? How does that sound to others?

Maybe I can re-title _Accessibility testing baseline_ to _How to test accessibility in Django_, and that will be the clearer beginner-friendly how-to?

carltongibson commented:
@sabderemane's post last week of her Accessibility checklist was right in the ballpark of what I think we need here. 

https://sarahabd.com/blog/my-accessibility-review-checklist

I think we could either link to that or, if Sarah were OK with it, bring it in.

thibaudcolas commented:
@nessita @carltongibson I’d appreciate your thoughts on my last comment first, before we consider adding standalone "how-to" content as separate pages.

If we can agree about the single-page "contributor guidelines" model, with more beginner-friendly material and a "how-to test" section, I’ll work on another draft. If you think what we really need first and foremost is a how-to guide or checklist on accessibility in the Django docs, I’ll close the PR and let other contributors have a go at _that_

carltongibson commented:
@thibaudcolas **for me** (note emphasis) I like what you have here. My only addition would be a note at the top roughly saying, "If you're looking to get starting implementing these guidelines, see this How To/Tutorial over here".

nessita commented:
> @nessita @carltongibson I’d appreciate your thoughts on my last comment first, before we consider adding standalone "how-to" content as separate pages.
> 
> If we can agree about the current single-page "contributor guidelines" model, with more beginner-friendly material and a "how-to test" section, I’ll work on another draft.

I think this is a good compromise. As Carlton said, with the recent article from Sarah A (which is so super helpful) and a smoother journey in the page from beginner friendly at the top -> more concrete and specific at the bottom, plus the now available Discord channel and Forum category, this would be a nice addition to our docs.

@sarahboyce given your recent new role as fellow, would you have some time to review this proposal, perhaps after @thibaudcolas pushes a new revision, so you can share your view as well?

thibaudcolas commented:
Cool! On it

sarahboyce commented on docs/internals/contributing/writing-code/accessibility.txt:
```diff
@@ -0,0 +1,124 @@
+Assistive technologies
+----------------------
+
+Where the UI changes could affect assistive technologies, here are popular
+free options we recommend testing with:
+
+* `NVDA`_ on Windows with Firefox ESR
+* `VoiceOver`_ on macOS with Safari
+* `Orca`_ on Linux with Firefox ESR
+* `Narrator`_ on Windows with Microsoft Edge
+* Windows Magnifier and macOS Zoom
+* Windows Speech Recognition and macOS Voice Control
+* VoiceOver on iOS, or TalkBack on Android
+* `Contrast themes`_ on Windows
```
```suggestion
Where the UI changes could affect assistive technologies, here are popular free
options we recommend testing with.

For Windows:

* `NVDA`_ - recommended with Firefox ESR
* `Narrator`_ - recommended with Microsoft Edge
* Windows Magnifier
* Windows Speech Recognition
* `Contrast themes`_

For macOS:

* `VoiceOver`_ - recommended with Safari
* macOS Zoom
* macOS Voice Control

For Linux:

* `Orca`_ - recommended with Firefox ESR

For mobile or tablet:

* VoiceOver on iOS, or TalkBack on Android
```
Suggestion: the reader will probably only have access to one OS and so I would group this and I tried to incorporate [this comment](https://github.com/django/django/pull/17338/files#r1487963102)
I am not sure on the "mobile or tablet" group name

sarahboyce reviewed (commented):
Thank you for this @thibaudcolas and all reviewers :heart: 

I think this is really close, I have a couple of suggestions adding a link or two and changes some wording but they are hopefully quite small things

sarahboyce commented:
Thank you all. I have pushed up some minor tweaks, @nessita can you check your happy with the recent changes?

nessita reviewed (approved):
Thank you @sarahboyce for picking this up, I think this is in great shape! :star2: 

I added some comments, nothing blocking.
